File: tools4a-view/src/main/java/com/github/panpf/tools4a/view/ViewAnimx.java
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.tools4a.view;

import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.animation.CycleInterpolator;
import android.view.animation.TranslateAnimation;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * View animation tool method
 */
public class ViewAnimx {

    /**
     * Default animation duration
     */
    public static final long DEFAULT_ANIMATION_DURATION = 400;


    private ViewAnimx() {
    }


    /**
     * Perform alpha animation
     *
     * @param fromAlpha      Initial alpha value
     * @param toAlpha        End alpha value
     * @param durationMillis Animation duration, in milliseconds
     * @param isBanClick     If true, clicks will be disabled during the animation
     * @param listener       Animation listener
     */
    public static void animAlpha(@NonNull final View view, float fromAlpha, float toAlpha, long durationMillis,
                                 final boolean isBanClick, @Nullable final Animation.AnimationListener listener) {
        view.clearAnimation();
        final boolean isClickable = view.isClickable();

        final AlphaAnimation anim = new AlphaAnimation(fromAlpha, toAlpha);
        anim.setDuration(durationMillis);
        anim.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(@NonNull Animation anim) {
                if (isClickable && isBanClick) view.setClickable(false);
                if (listener != null) listener.onAnimationStart(anim);
            }

            @Override
            public void onAnimationRepeat(@NonNull Animation anim) {
                if (listener != null) listener.onAnimationRepeat(anim);
            }

            @Override
            public void onAnimationEnd(@NonNull Animation anim) {
                if (isClickable && isBanClick) view.setClickable(true);
                if (listener != null) listener.onAnimationEnd(anim);
            }
        });
        view.startAnimation(anim);
    }

    /**
     * Perform alpha animation
     *
     * @param fromAlpha Initial alpha value
     * @param toAlpha   End alpha value
     * @param listener  Animation listener
     */
    public static void animAlpha(@NonNull final View view, float fromAlpha, float toAlpha, @Nullable final Animation.AnimationListener listener) {
        animAlpha(view, fromAlpha, toAlpha, DEFAULT_ANIMATION_DURATION, false, listener);
    }

    /**
     * Perform alpha animation
     *
     * @param fromAlpha  Initial alpha value
     * @param toAlpha    End alpha value
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void animAlpha(@NonNull final View view, float fromAlpha, float toAlpha, boolean isBanClick) {
        animAlpha(view, fromAlpha, toAlpha, DEFAULT_ANIMATION_DURATION, isBanClick, null);
    }

    /**
     * Perform alpha animation
     *
     * @param fromAlpha      Initial alpha value
     * @param toAlpha        End alpha value
     * @param durationMillis Animation duration, in milliseconds
     */
    public static void animAlpha(@NonNull final View view, float fromAlpha, float toAlpha, long durationMillis) {
        animAlpha(view, fromAlpha, toAlpha, durationMillis, false, null);
    }

    /**
     * Perform alpha animation
     *
     * @param fromAlpha Initial alpha value
     * @param toAlpha   End alpha value
     */
    public static void animAlpha(@NonNull final View view, float fromAlpha, float toAlpha) {
        animAlpha(view, fromAlpha, toAlpha, DEFAULT_ANIMATION_DURATION, false, null);
    }


    /**
     * Perform translate animation
     *
     * @param fromXDelta     Change in X coordinate to apply at the start of the animation
     * @param toXDelta       Change in X coordinate to apply at the end of the animation
     * @param fromYDelta     Change in Y coordinate to apply at the start of the animation
     * @param toYDelta       Change in Y coordinate to apply at the end of the animation
     * @param cycles         Number of cycle executions
     * @param durationMillis Animation duration, in milliseconds
     * @param isBanClick     If true, clicks will be disabled during the animation
     * @param listener       Animation listener
     */
    public static void animTranslate(@NonNull final View view, float fromXDelta, float toXDelta, float fromYDelta, float toYDelta,
                                     float cycles, long durationMillis, final boolean isBanClick, @Nullable final Animation.AnimationListener listener) {
        view.clearAnimation();
        final boolean isClickable = view.isClickable();

        TranslateAnimation anim = new TranslateAnimation(fromXDelta, toXDelta, fromYDelta, toYDelta);
        anim.setDuration(durationMillis);
        if (cycles > 0.0f) {
            anim.setInterpolator(new CycleInterpolator(cycles));
        }
        anim.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(@NonNull Animation anim) {
                if (isClickable && isBanClick) view.setClickable(false);
                if (listener != null) listener.onAnimationStart(anim);
            }

            @Override
            public void onAnimationRepeat(@NonNull Animation anim) {
                if (listener != null) listener.onAnimationRepeat(anim);
            }

            @Override
            public void onAnimationEnd(@NonNull Animation anim) {
                if (isClickable && isBanClick) view.setClickable(true);
                if (listener != null) listener.onAnimationEnd(anim);
            }
        });
        view.startAnimation(anim);
    }

    /**
     * Perform translate animation
     *
     * @param fromXDelta Change in X coordinate to apply at the start of the animation
     * @param toXDelta   Change in X coordinate to apply at the end of the animation
     * @param fromYDelta Change in Y coordinate to apply at the start of the animation
     * @param toYDelta   Change in Y coordinate to apply at the end of the animation
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void animTranslate(@NonNull final View view, float fromXDelta, float toXDelta, float fromYDelta, float toYDelta, final boolean isBanClick) {
        animTranslate(view, fromXDelta, toXDelta, fromYDelta, toYDelta, 0, DEFAULT_ANIMATION_DURATION, isBanClick, null);
    }

    /**
     * Perform translate animation
     *
     * @param fromXDelta Change in X coordinate to apply at the start of the animation
     * @param toXDelta   Change in X coordinate to apply at the end of the animation
     * @param fromYDelta Change in Y coordinate to apply at the start of the animation
     * @param toYDelta   Change in Y coordinate to apply at the end of the animation
     * @param listener   Animation listener
     */
    public static void animTranslate(@NonNull final View view, float fromXDelta, float toXDelta, float fromYDelta, float toYDelta, @Nullable final Animation.AnimationListener listener) {
        animTranslate(view, fromXDelta, toXDelta, fromYDelta, toYDelta, 0, DEFAULT_ANIMATION_DURATION, false, listener);
    }

    /**
     * Perform translate animation
     *
     * @param fromXDelta     Change in X coordinate to apply at the start of the animation
     * @param toXDelta       Change in X coordinate to apply at the end of the animation
     * @param fromYDelta     Change in Y coordinate to apply at the start of the animation
     * @param toYDelta       Change in Y coordinate to apply at the end of the animation
     * @param durationMillis Animation duration, in milliseconds
     */
    public static void animTranslate(@NonNull final View view, float fromXDelta, float toXDelta, float fromYDelta, float toYDelta, long durationMillis) {
        animTranslate(view, fromXDelta, toXDelta, fromYDelta, toYDelta, 0, durationMillis, false, null);
    }

    /**
     * Perform translate animation
     *
     * @param fromXDelta Change in X coordinate to apply at the start of the animation
     * @param toXDelta   Change in X coordinate to apply at the end of the animation
     * @param fromYDelta Change in Y coordinate to apply at the start of the animation
     * @param toYDelta   Change in Y coordinate to apply at the end of the animation
     * @param cycles     Number of cycle executions
     */
    public static void animTranslate(@NonNull final View view, float fromXDelta, float toXDelta, float fromYDelta, float toYDelta,
                                     float cycles) {
        animTranslate(view, fromXDelta, toXDelta, fromYDelta, toYDelta, cycles, DEFAULT_ANIMATION_DURATION, false, null);
    }

    /**
     * Perform translate animation
     *
     * @param fromXDelta Change in X coordinate to apply at the start of the animation
     * @param toXDelta   Change in X coordinate to apply at the end of the animation
     * @param fromYDelta Change in Y coordinate to apply at the start of the animation
     * @param toYDelta   Change in Y coordinate to apply at the end of the animation
     */
    public static void animTranslate(@NonNull final View view, float fromXDelta, float toXDelta, float fromYDelta, float toYDelta) {
        animTranslate(view, fromXDelta, toXDelta, fromYDelta, toYDelta, 0, DEFAULT_ANIMATION_DURATION, false, null);
    }


    /**
     * Shake left and right
     *
     * @param extent         Shake amplitude
     * @param cycles         Number of cycle executions
     * @param durationMillis Animation duration, in milliseconds
     * @param isBanClick     If true, clicks will be disabled during the animation
     * @param listener       Animation listener
     */
    public static void shakeLandscape(@NonNull final View view, float extent, float cycles, long durationMillis,
                                      boolean isBanClick, @Nullable final Animation.AnimationListener listener) {
        animTranslate(view, 0.0f, extent, 0.0f, 0.0f, cycles, durationMillis, isBanClick, listener);
    }

    /**
     * Shake left and right
     *
     * @param listener Animation listener
     */
    public static void shakeLandscape(@NonNull final View view, @Nullable final Animation.AnimationListener listener) {
        shakeLandscape(view, 10f, 7f, 700, false, listener);
    }

    /**
     * Shake left and right
     *
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void shakeLandscape(@NonNull final View view, boolean isBanClick) {
        shakeLandscape(view, 10f, 7f, 700, isBanClick, null);
    }

    /**
     * Shake left and right
     *
     * @param durationMillis Animation duration, in milliseconds
     */
    public static void shakeLandscape(@NonNull final View view, long durationMillis) {
        shakeLandscape(view, 10f, 7f, durationMillis, false, null);
    }

    /**
     * Shake left and right
     *
     * @param extent Shake amplitude
     */
    public static void shakeLandscape(@NonNull final View view, float extent) {
        shakeLandscape(view, extent, 7f, 700, false, null);
    }

    /**
     * Shake left and right
     */
    public static void shakeLandscape(@NonNull final View view) {
        shakeLandscape(view, 10f, 7, 700, false, null);
    }


    /**
     * Shock up and down
     *
     * @param extent         Shake amplitude
     * @param cycles         Number of cycle executions
     * @param durationMillis Animation duration, in milliseconds
     * @param isBanClick     If true, clicks will be disabled during the animation
     * @param listener       Animation listener
     */
    public static void shakePortrait(@NonNull final View view, float extent, float cycles, long durationMillis,
                                     boolean isBanClick, @Nullable final Animation.AnimationListener listener) {
        animTranslate(view, 0.0f, 0.0f, 0.0f, extent, cycles, durationMillis, isBanClick, listener);
    }

    /**
     * Shock up and down
     *
     * @param listener Animation listener
     */
    public static void shakePortrait(@NonNull final View view, @Nullable final Animation.AnimationListener listener) {
        shakePortrait(view, 10.f, 7f, 700, false, listener);
    }

    /**
     * Shock up and down
     *
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void shakePortrait(@NonNull final View view, boolean isBanClick) {
        shakePortrait(view, 10.f, 7f, 700, isBanClick, null);
    }

    /**
     * Shock up and down
     *
     * @param durationMillis Animation duration, in milliseconds
     */
    public static void shakePortrait(@NonNull final View view, long durationMillis) {
        shakePortrait(view, 10.f, 7f, durationMillis, false, null);
    }

    /**
     * Shock up and down
     *
     * @param extent Shake amplitude
     */
    public static void shakePortrait(@NonNull final View view, float extent) {
        shakePortrait(view, extent, 7f, 700, false, null);
    }

    /**
     * Shock up and down
     */
    public static void shakePortrait(@NonNull final View view) {
        shakePortrait(view, 10.f, 7f, 700, false, null);
    }


    /**
     * Parse the animation from the resource file and execute
     *
     * @param animId     Animation resource ID
     * @param isBanClick If true, clicks will be disabled during the animation
     * @param listener   Animation listener
     */
    public static void startAnimFromRes(@NonNull final View view, int animId, final boolean isBanClick,
                                        @Nullable final Animation.AnimationListener listener) {
        view.clearAnimation();
        final boolean isClickable = view.isClickable();

        Animation anim = AnimationUtils.loadAnimation(view.getContext(), animId);
        anim.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(@NonNull Animation anim) {
                if (isClickable && isBanClick) view.setClickable(false);
                if (listener != null) listener.onAnimationStart(anim);
            }

            @Override
            public void onAnimationRepeat(@NonNull Animation anim) {
                if (listener != null) listener.onAnimationRepeat(anim);
            }

            @Override
            public void onAnimationEnd(@NonNull Animation anim) {
                if (isClickable && isBanClick) view.setClickable(true);
                if (listener != null) listener.onAnimationEnd(anim);
            }
        });
        view.startAnimation(anim);
    }

    /**
     * Parse the animation from the resource file and execute
     *
     * @param animId   Animation resource ID
     * @param listener Animation listener
     */
    public static void startAnimFromRes(@NonNull final View view, int animId, @Nullable final Animation.AnimationListener listener) {
        startAnimFromRes(view, animId, false, listener);
    }

    /**
     * Parse the animation from the resource file and execute
     *
     * @param animId     Animation resource ID
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void startAnimFromRes(@NonNull final View view, int animId, final boolean isBanClick) {
        startAnimFromRes(view, animId, isBanClick, null);
    }

    /**
     * Parse the animation from the resource file and execute
     *
     * @param animId Animation resource ID
     */
    public static void startAnimFromRes(@NonNull final View view, int animId) {
        startAnimFromRes(view, animId, false, null);
    }


    /**
     * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
     *
     * @param durationMillis Animation duration, in milliseconds
     * @param isBanClick     If true, clicks will be disabled during the animation
     * @param listener       Animation listener
     */
    public static void invisibleByAnimAlpha(@NonNull final View view, long durationMillis,
                                            final boolean isBanClick, @Nullable final Animation.AnimationListener listener) {
        if (view.getVisibility() == View.INVISIBLE) return;
        animAlpha(view, 1.0f, 0.0f, durationMillis, isBanClick, new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(@NonNull Animation anim) {
                if (listener != null) listener.onAnimationStart(anim);
            }

            @Override
            public void onAnimationRepeat(@NonNull Animation anim) {
                if (listener != null) listener.onAnimationRepeat(anim);
            }

            @Override
            public void onAnimationEnd(@NonNull Animation anim) {
                view.setVisibility(View.INVISIBLE);
                if (listener != null) listener.onAnimationEnd(anim);
            }
        });
    }

    /**
     * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
     *
     * @param listener Animation listener
     */
    public static void invisibleByAnimAlpha(@NonNull final View view, @Nullable final Animation.AnimationListener listener) {
        invisibleByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, false, listener);
    }

    /**
     * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
     *
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void invisibleByAnimAlpha(@NonNull final View view, final boolean isBanClick) {
        invisibleByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, isBanClick, null);
    }

    /**
     * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
     *
     * @param durationMillis Animation duration, in milliseconds
     */
    public static void invisibleByAnimAlpha(@NonNull final View view, long durationMillis) {
        invisibleByAnimAlpha(view, durationMillis, false, null);
    }

    /**
     * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
     */
    public static void invisibleByAnimAlpha(@NonNull final View view) {
        invisibleByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, false, null);
    }


    /**
     * Hide the view with a alpha animation, with visibility set to GONE at the end
     *
     * @param durationMillis Animation duration, in milliseconds
     * @param isBanClick     If true, clicks will be disabled during the animation
     * @param listener       Animation listener
     */
    public static void goneByAnimAlpha(@NonNull final View view, long durationMillis,
                                       final boolean isBanClick, @Nullable final Animation.AnimationListener listener) {
        if (view.getVisibility() == View.GONE) return;
        animAlpha(view, 1.0f, 0.0f, durationMillis, isBanClick, new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(@NonNull Animation anim) {
                if (listener != null) listener.onAnimationStart(anim);
            }

            @Override
            public void onAnimationRepeat(@NonNull Animation anim) {
                if (listener != null) listener.onAnimationRepeat(anim);
            }

            @Override
            public void onAnimationEnd(@NonNull Animation anim) {
                view.setVisibility(View.GONE);
                if (listener != null) listener.onAnimationEnd(anim);
            }
        });
    }

    /**
     * Hide the view with a alpha animation, with visibility set to GONE at the end
     *
     * @param listener Animation listener
     */
    public static void goneByAnimAlpha(@NonNull final View view, @Nullable final Animation.AnimationListener listener) {
        goneByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, false, listener);
    }

    /**
     * Hide the view with a alpha animation, with visibility set to GONE at the end
     *
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void goneByAnimAlpha(@NonNull final View view, final boolean isBanClick) {
        goneByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, isBanClick, null);
    }

    /**
     * Hide the view with a alpha animation, with visibility set to GONE at the end
     *
     * @param durationMillis Animation duration, in milliseconds
     */
    public static void goneByAnimAlpha(@NonNull final View view, long durationMillis) {
        goneByAnimAlpha(view, durationMillis, false, null);
    }

    /**
     * Hide the view with a alpha animation, with visibility set to GONE at the end
     */
    public static void goneByAnimAlpha(@NonNull final View view) {
        goneByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, false, null);
    }


    /**
     * Use alpha animation to display View
     *
     * @param durationMillis Animation duration, in milliseconds
     * @param isBanClick     If true, clicks will be disabled during the animation
     * @param listener       Animation listener
     */
    public static void visibleByAnimAlpha(@NonNull final View view, long durationMillis,
                                          final boolean isBanClick, @Nullable final Animation.AnimationListener listener) {
        if (view.getVisibility() == View.VISIBLE) return;
        view.setVisibility(View.VISIBLE);
        animAlpha(view, 0.0f, 1.0f, durationMillis, isBanClick, listener);
    }

    /**
     * Use alpha animation to display View
     *
     * @param listener Animation listener
     */
    public static void visibleByAnimAlpha(@NonNull final View view, @Nullable final Animation.AnimationListener listener) {
        visibleByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, false, listener);
    }

    /**
     * Use alpha animation to display View
     *
     * @param isBanClick If true, clicks will be disabled during the animation
     */
    public static void visibleByAnimAlpha(@NonNull final View view, final boolean isBanClick) {
        visibleByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, isBanClick, null);
    }

    /**
     * Use alpha animation to display View
     *
     * @param durationMillis Animation duration, in milliseconds
     */
    public static void visibleByAnimAlpha(@NonNull final View view, long durationMillis) {
        visibleByAnimAlpha(view, durationMillis, false, null);
    }

    /**
     * Use alpha animation to display View
     */
    public static void visibleByAnimAlpha(@NonNull final View view) {
        visibleByAnimAlpha(view, DEFAULT_ANIMATION_DURATION, false, null);
    }
}

File: tools4a-view-ktx/src/main/java/com/github/panpf/tools4a/view/ktx/ViewAnimx.kt
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("NOTHING_TO_INLINE")

package com.github.panpf.tools4a.view.ktx

import android.view.View
import android.view.animation.Animation
import com.github.panpf.tools4a.view.ViewAnimx


/*
 * View animation related extension methods or properties
 */


/**
 * Perform alpha animation
 *
 * @param fromAlpha      Initial alpha value
 * @param toAlpha        End alpha value
 * @param durationMillis Animation duration, in milliseconds
 * @param isBanClick     If true, clicks will be disabled during the animation
 * @param listener       Animation listener
 */
inline fun View.animAlpha(fromAlpha: Float, toAlpha: Float, durationMillis: Long, isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.animAlpha(this, fromAlpha, toAlpha, durationMillis, isBanClick, listener)

/**
 * Perform alpha animation
 *
 * @param fromAlpha Initial alpha value
 * @param toAlpha   End alpha value
 * @param listener  Animation listener
 */
inline fun View.animAlpha(fromAlpha: Float, toAlpha: Float, listener: Animation.AnimationListener?) =
        ViewAnimx.animAlpha(this, fromAlpha, toAlpha, listener)

/**
 * Perform alpha animation
 *
 * @param fromAlpha  Initial alpha value
 * @param toAlpha    End alpha value
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.animAlpha(fromAlpha: Float, toAlpha: Float, isBanClick: Boolean) = ViewAnimx.animAlpha(this, fromAlpha, toAlpha, isBanClick)

/**
 * Perform alpha animation
 *
 * @param fromAlpha      Initial alpha value
 * @param toAlpha        End alpha value
 * @param durationMillis Animation duration, in milliseconds
 */
inline fun View.animAlpha(fromAlpha: Float, toAlpha: Float, durationMillis: Long) = ViewAnimx.animAlpha(this, fromAlpha, toAlpha, durationMillis)

/**
 * Perform alpha animation
 *
 * @param fromAlpha Initial alpha value
 * @param toAlpha   End alpha value
 */
inline fun View.animAlpha(fromAlpha: Float, toAlpha: Float) = ViewAnimx.animAlpha(this, fromAlpha, toAlpha)


/**
 * Perform translate animation
 *
 * @param fromXDelta     Change in X coordinate to apply at the start of the animation
 * @param toXDelta       Change in X coordinate to apply at the end of the animation
 * @param fromYDelta     Change in Y coordinate to apply at the start of the animation
 * @param toYDelta       Change in Y coordinate to apply at the end of the animation
 * @param cycles         Number of cycle executions
 * @param durationMillis Animation duration, in milliseconds
 * @param isBanClick     If true, clicks will be disabled during the animation
 * @param listener       Animation listener
 */
inline fun View.animTranslate(fromXDelta: Float, toXDelta: Float, fromYDelta: Float, toYDelta: Float,
                              cycles: Float, durationMillis: Long, isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.animTranslate(this, fromXDelta, toXDelta, fromYDelta, toYDelta, cycles, durationMillis, isBanClick, listener)

/**
 * Perform translate animation
 *
 * @param fromXDelta Change in X coordinate to apply at the start of the animation
 * @param toXDelta   Change in X coordinate to apply at the end of the animation
 * @param fromYDelta Change in Y coordinate to apply at the start of the animation
 * @param toYDelta   Change in Y coordinate to apply at the end of the animation
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.animTranslate(fromXDelta: Float, toXDelta: Float, fromYDelta: Float, toYDelta: Float, isBanClick: Boolean) =
        ViewAnimx.animTranslate(this, fromXDelta, toXDelta, fromYDelta, toYDelta, isBanClick)

/**
 * Perform translate animation
 *
 * @param fromXDelta Change in X coordinate to apply at the start of the animation
 * @param toXDelta   Change in X coordinate to apply at the end of the animation
 * @param fromYDelta Change in Y coordinate to apply at the start of the animation
 * @param toYDelta   Change in Y coordinate to apply at the end of the animation
 * @param listener   Animation listener
 */
inline fun View.animTranslate(fromXDelta: Float, toXDelta: Float, fromYDelta: Float, toYDelta: Float, listener: Animation.AnimationListener?) =
        ViewAnimx.animTranslate(this, fromXDelta, toXDelta, fromYDelta, toYDelta, listener)

/**
 * Perform translate animation
 *
 * @param fromXDelta     Change in X coordinate to apply at the start of the animation
 * @param toXDelta       Change in X coordinate to apply at the end of the animation
 * @param fromYDelta     Change in Y coordinate to apply at the start of the animation
 * @param toYDelta       Change in Y coordinate to apply at the end of the animation
 * @param durationMillis Animation duration, in milliseconds
 */
inline fun View.animTranslate(fromXDelta: Float, toXDelta: Float, fromYDelta: Float, toYDelta: Float, durationMillis: Long) =
        ViewAnimx.animTranslate(this, fromXDelta, toXDelta, fromYDelta, toYDelta, durationMillis)

/**
 * Perform translate animation
 *
 * @param fromXDelta Change in X coordinate to apply at the start of the animation
 * @param toXDelta   Change in X coordinate to apply at the end of the animation
 * @param fromYDelta Change in Y coordinate to apply at the start of the animation
 * @param toYDelta   Change in Y coordinate to apply at the end of the animation
 * @param cycles     Number of cycle executions
 */
inline fun View.animTranslate(fromXDelta: Float, toXDelta: Float, fromYDelta: Float, toYDelta: Float, cycles: Float) =
        ViewAnimx.animTranslate(this, fromXDelta, toXDelta, fromYDelta, toYDelta, cycles)

/**
 * Perform translate animation
 *
 * @param fromXDelta Change in X coordinate to apply at the start of the animation
 * @param toXDelta   Change in X coordinate to apply at the end of the animation
 * @param fromYDelta Change in Y coordinate to apply at the start of the animation
 * @param toYDelta   Change in Y coordinate to apply at the end of the animation
 */
inline fun View.animTranslate(fromXDelta: Float, toXDelta: Float, fromYDelta: Float, toYDelta: Float) =
        ViewAnimx.animTranslate(this, fromXDelta, toXDelta, fromYDelta, toYDelta)


/**
 * Shake left and right
 *
 * @param extent         Shake amplitude
 * @param cycles         Number of cycle executions
 * @param durationMillis Animation duration, in milliseconds
 * @param isBanClick     If true, clicks will be disabled during the animation
 * @param listener       Animation listener
 */
inline fun View.shakeLandscape(extent: Float, cycles: Float, durationMillis: Long,
                               isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.shakeLandscape(this, extent, cycles, durationMillis, isBanClick, listener)

/**
 * Shake left and right
 *
 * @param listener Animation listener
 */
inline fun View.shakeLandscape(listener: Animation.AnimationListener?) = ViewAnimx.shakeLandscape(this, listener)

/**
 * Shake left and right
 *
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.shakeLandscape(isBanClick: Boolean) = ViewAnimx.shakeLandscape(this, isBanClick)

/**
 * Shake left and right
 *
 * @param durationMillis Animation duration, in milliseconds
 */
inline fun View.shakeLandscape(durationMillis: Long) = ViewAnimx.shakeLandscape(this, durationMillis)

/**
 * Shake left and right
 *
 * @param extent Shake amplitude
 */
inline fun View.shakeLandscape(extent: Float) = ViewAnimx.shakeLandscape(this, extent)

/**
 * Shake left and right
 */
inline fun View.shakeLandscape() = ViewAnimx.shakeLandscape(this)


/**
 * Shock up and down
 *
 * @param extent         Shake amplitude
 * @param cycles         Number of cycle executions
 * @param durationMillis Animation duration, in milliseconds
 * @param isBanClick     If true, clicks will be disabled during the animation
 * @param listener       Animation listener
 */
inline fun View.shakePortrait(extent: Float, cycles: Float, durationMillis: Long, isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.shakePortrait(this, extent, cycles, durationMillis, isBanClick, listener)

/**
 * Shock up and down
 *
 * @param listener Animation listener
 */
inline fun View.shakePortrait(listener: Animation.AnimationListener?) = ViewAnimx.shakePortrait(this, listener)

/**
 * Shock up and down
 *
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.shakePortrait(isBanClick: Boolean) = ViewAnimx.shakePortrait(this, isBanClick)

/**
 * Shock up and down
 *
 * @param durationMillis Animation duration, in milliseconds
 */
inline fun View.shakePortrait(durationMillis: Long) = ViewAnimx.shakePortrait(this, durationMillis)

/**
 * Shock up and down
 *
 * @param extent Shake amplitude
 */
inline fun View.shakePortrait(extent: Float) = ViewAnimx.shakePortrait(this, extent)

/**
 * Shock up and down
 */
inline fun View.shakePortrait() = ViewAnimx.shakePortrait(this)


/**
 * Parse the animation from the resource file and execute
 *
 * @param animId     Animation resource ID
 * @param isBanClick If true, clicks will be disabled during the animation
 * @param listener   Animation listener
 */
inline fun View.startAnimFromRes(animId: Int, isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.startAnimFromRes(this, animId, isBanClick, listener)

/**
 * Parse the animation from the resource file and execute
 *
 * @param animId   Animation resource ID
 * @param listener Animation listener
 */
inline fun View.startAnimFromRes(animId: Int, listener: Animation.AnimationListener?) = ViewAnimx.startAnimFromRes(this, animId, listener)

/**
 * Parse the animation from the resource file and execute
 *
 * @param animId     Animation resource ID
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.startAnimFromRes(animId: Int, isBanClick: Boolean) = ViewAnimx.startAnimFromRes(this, animId, isBanClick)

/**
 * Parse the animation from the resource file and execute
 *
 * @param animId Animation resource ID
 */
inline fun View.startAnimFromRes(animId: Int) = ViewAnimx.startAnimFromRes(this, animId)


/**
 * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
 *
 * @param durationMillis Animation duration, in milliseconds
 * @param isBanClick     If true, clicks will be disabled during the animation
 * @param listener       Animation listener
 */
inline fun View.invisibleByAnimAlpha(durationMillis: Long, isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.invisibleByAnimAlpha(this, durationMillis, isBanClick, listener)

/**
 * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
 *
 * @param listener Animation listener
 */
inline fun View.invisibleByAnimAlpha(listener: Animation.AnimationListener?) = ViewAnimx.invisibleByAnimAlpha(this, listener)

/**
 * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
 *
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.invisibleByAnimAlpha(isBanClick: Boolean) = ViewAnimx.invisibleByAnimAlpha(this, isBanClick)

/**
 * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
 *
 * @param durationMillis Animation duration, in milliseconds
 */
inline fun View.invisibleByAnimAlpha(durationMillis: Long) = ViewAnimx.invisibleByAnimAlpha(this, durationMillis)

/**
 * Hide the view with a alpha animation, with visibility set to INVISIBLE at the end
 */
inline fun View.invisibleByAnimAlpha() = ViewAnimx.invisibleByAnimAlpha(this)


/**
 * Hide the view with a alpha animation, with visibility set to GONE at the end
 *
 * @param durationMillis Animation duration, in milliseconds
 * @param isBanClick     If true, clicks will be disabled during the animation
 * @param listener       Animation listener
 */
inline fun View.goneByAnimAlpha(durationMillis: Long, isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.goneByAnimAlpha(this, durationMillis, isBanClick, listener)

/**
 * Hide the view with a alpha animation, with visibility set to GONE at the end
 *
 * @param listener Animation listener
 */
inline fun View.goneByAnimAlpha(listener: Animation.AnimationListener?) = ViewAnimx.goneByAnimAlpha(this, listener)

/**
 * Hide the view with a alpha animation, with visibility set to GONE at the end
 *
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.goneByAnimAlpha(isBanClick: Boolean) = ViewAnimx.goneByAnimAlpha(this, isBanClick)

/**
 * Hide the view with a alpha animation, with visibility set to GONE at the end
 *
 * @param durationMillis Animation duration, in milliseconds
 */
inline fun View.goneByAnimAlpha(durationMillis: Long) = ViewAnimx.goneByAnimAlpha(this, durationMillis)

/**
 * Hide the view with a alpha animation, with visibility set to GONE at the end
 */
inline fun View.goneByAnimAlpha() = ViewAnimx.goneByAnimAlpha(this)


/**
 * Use alpha animation to display View
 *
 * @param durationMillis Animation duration, in milliseconds
 * @param isBanClick     If true, clicks will be disabled during the animation
 * @param listener       Animation listener
 */
inline fun View.visibleByAnimAlpha(durationMillis: Long, isBanClick: Boolean, listener: Animation.AnimationListener?) =
        ViewAnimx.visibleByAnimAlpha(this, durationMillis, isBanClick, listener)

/**
 * Use alpha animation to display View
 *
 * @param listener Animation listener
 */
inline fun View.visibleByAnimAlpha(listener: Animation.AnimationListener?) = ViewAnimx.visibleByAnimAlpha(this, listener)

/**
 * Use alpha animation to display View
 *
 * @param isBanClick If true, clicks will be disabled during the animation
 */
inline fun View.visibleByAnimAlpha(isBanClick: Boolean) = ViewAnimx.visibleByAnimAlpha(this, isBanClick)

/**
 * Use alpha animation to display View
 *
 * @param durationMillis Animation duration, in milliseconds
 */
inline fun View.visibleByAnimAlpha(durationMillis: Long) = ViewAnimx.visibleByAnimAlpha(this, durationMillis)

/**
 * Use alpha animation to display View
 */
inline fun View.visibleByAnimAlpha() = ViewAnimx.visibleByAnimAlpha(this)
